Broken upgrader in `SwordUpgradeTest`

The failure is in `axiom.test.test_upgrading.SwordUpgradeTest.testUpgradeWithMissingVersion`, but it only occurs when a certain dict ordering happens. `PYTHONHASHSEED=1290646218` produces one such ordering; this wasn't noticed before because the default seed doesn't produce the ordering that causes the failure.

First, the failing test traceback:
```
[ERROR]
Traceback (most recent call last):
  File "/home/mithrandi/code/axiom/axiom/test/test_upgrading.py", line 189, in afterUpgrade
    self._testPlayerAndSwordState(player, sword)
  File "/home/mithrandi/code/axiom/axiom/test/test_upgrading.py", line 368, in _testPlayerAndSwordState
    assert not sword.__legacy__
exceptions.AssertionError: 

axiom.test.test_upgrading.SwordUpgradeTest.testUpgradeWithMissingVersion
```
The assertion fires because the sword item has not been fully upgraded. The reason for this is that the `player1to2` upgrader does something that confuses the upgrade system. When the store is opened, Axiom looks at all of the currently-existing type versions in the store; any version that is out of date is handed to the upgrade manager in order to be upgraded. In the case of this test, the versions are `(Player, 1)` and `(Sword, 1)`. The upgrade manager loops over the types to be upgraded; for each type, it queries for all items of that type, and fully upgrades them (running upgraders until the current schema version is reached), before moving on to the next type/version.

The `player1to2` upgrader calls `upgradePlayerAndSword`, which retrieves both the sword and the player, and upgrades both of them to version 2. If this upgrader runs first, then the sword item will be at version 2. However, recall that the entry in the upgrader's list is `(Sword, 1)`; when it queries for items of this type, it obviously finds none. However, `(Sword, 2)` is not in its list, so the sword version 2 item is never upgraded to sword version 3. (The inverse problem does not occur when `sword1to2` runs first, because there is no Player v3 type).

How to proceed? I'm not sure if what this upgrader is doing should just be considered unsupported, and this upgrader (and corresponding part of the test) should be removed, or if we should find some way to make this functionality actually work (probably not impossible, but I don't see an easy way to make it work at the moment, so likely quite a lot of work).

--- a/axiom/attributes.py
+++ b/axiom/attributes.py
@@ -1123,7 +1123,7 @@ class reference(integer):
             assert self.whenDeleted is reference.NULLIFY, (
                 "not sure what to do if not...")
             return None
-        if rv.__legacy__:
+        if rv.__legacy__ and not oself.__legacy__:
             delattr(oself, self.underlying)
             return super(reference, self).__get__(oself, cls)
         return rv

--- a/axiom/attributes.py
+++ b/axiom/attributes.py
@@ -1123,6 +1123,11 @@ class reference(integer):
             assert self.whenDeleted is reference.NULLIFY, (
                 "not sure what to do if not...")
             return None
+        # If the current cached value is a legacy item, we discard it in order
+        # to force another fetch from the database. However, if *this item* is
+        # also a legacy item, then the item referred to may have been created
+        # in an upgrader and not have been stored yet, so we shouldn't discard
+        # it.
         if rv.__legacy__ and not oself.__legacy__:
             delattr(oself, self.underlying)
             return super(reference, self).__get__(oself, cls)

--- a/axiom/store.py
+++ b/axiom/store.py
@@ -2099,6 +2099,10 @@ class Store(Empowered):
         if self.tablesCreatedThisTransaction is not None:
             self.tablesCreatedThisTransaction.append(tableClass)
 
+        cls = _typeNameToMostRecentClass.get(tableClass.typeName)
+        if cls is not None and tableClass.schemaVersion != cls.schemaVersion:
+            self._upgradeManager.queueTypeUpgrade(tableClass)
+
         # We can pass () for extantIndexes here because since the table didn't
         # exist for tableClass, none of its indexes could have either.
         # Whatever checks _createIndexesFor will make would give the same

--- a/axiom/store.py
+++ b/axiom/store.py
@@ -2099,6 +2099,9 @@ class Store(Empowered):
         if self.tablesCreatedThisTransaction is not None:
             self.tablesCreatedThisTransaction.append(tableClass)
 
+        # If the new type is a legacy type (not the current version), we need
+        # to queue it for upgrade to ensure that if we are in the middle of an
+        # upgrade, legacy items of this version get upgraded.
         cls = _typeNameToMostRecentClass.get(tableClass.typeName)
         if cls is not None and tableClass.schemaVersion != cls.schemaVersion:
             self._upgradeManager.queueTypeUpgrade(tableClass)

--- a/axiom/test/morenewapp.py
+++ b/axiom/test/morenewapp.py
@@ -4,7 +4,7 @@
 from axiom.item import Item
 from axiom.attributes import text, integer, reference, inmemory
 
-from axiom.upgrade import registerUpgrader
+from axiom.upgrade import registerUpgrader, registerAttributeCopyingUpgrader
 
 class ActivateHelper:
     activated = 0
@@ -13,7 +13,7 @@ class ActivateHelper:
 
 class Adventurer(ActivateHelper, Item):
     typeName = 'test_app_player'
-    schemaVersion = 2
+    schemaVersion = 3
 
     name = text()
     activated = inmemory()
@@ -64,6 +64,15 @@ registerUpgrader(sword2to3, 'test_app_sword', 2, 3)
 # declare legacy class.
 
 from axiom.item import declareLegacyItem
+
+
+declareLegacyItem(
+    typeName='test_app_player',
+    schemaVersion=2,
+    attributes=dict(
+        name=text(allowNone=True)))
+
+registerAttributeCopyingUpgrader(Adventurer, 2, 3)
 
 declareLegacyItem(typeName = 'test_app_sword',
                   schemaVersion = 2,

--- a/axiom/test/test_reference.py
+++ b/axiom/test/test_reference.py
@@ -3,7 +3,7 @@ import gc
 from twisted.trial.unittest import TestCase
 
 from axiom.store import Store
-from axiom.upgrade import registerUpgrader
+from axiom.upgrade import registerUpgrader, registerAttributeCopyingUpgrader
 from axiom.item import Item, declareLegacyItem
 from axiom.attributes import integer, reference
 from axiom.errors import BrokenReference, DeletionDisallowed
@@ -187,16 +187,33 @@ class BadReferenceTestCase(TestCase):
         """
         store = Store()
         referent = SimpleReferent(store=store)
-        oldReferee = nonUpgradedItem(store=store)
+        oldReferee = nonUpgradedItem2(store=store)
         referent.ref = oldReferee
         # Manually run the upgrader on this specific legacy item. This is the
         # same as if the SimpleReferent item had been created in an upgrader
         # for UpgradedItem, except that we can keep a strong reference to
         # oldReferee to ensure it is not garbage collected (this would
         # otherwise happen nondeterministically on platforms like PyPy).
-        newReferee = item1to2(oldReferee)
+        newReferee = item2to3(oldReferee)
         self.assertIsInstance(newReferee, UpgradedItem)
         self.assertIdentical(referent.ref, newReferee)
+
+
+    def test_dummyItemReferenceInUpgrade(self):
+        """
+        Setting the value of a reference attribute to a legacy item during an
+        upgrade does not lose the value.
+        """
+        store = Store()
+        def tx():
+            oldReferent = nonUpgradedItem(store=store)
+            oldReferee = nonUpgradedItem(store=store)
+            newReferent = oldReferent.upgradeVersion(
+                UpgradedItem.typeName, 1, 2)
+            newReferee = oldReferee.upgradeVersion(
+                UpgradedItem.typeName, 1, 2, ref=newReferent)
+            self.assertIdentical(newReferee.ref, newReferent)
+        store.transact(tx)
 
 
     def test_dummyItemGetItemByID(self):
@@ -218,21 +235,28 @@ class UpgradedItem(Item):
     """
     A simple item which is the current version of L{nonUpgradedItem}.
     """
-    schemaVersion = 2
-    dummy = integer()
+    schemaVersion = 3
+    ref = reference()
 
 
 
 nonUpgradedItem = declareLegacyItem(
     UpgradedItem.typeName, 1,
-    dict(dummy=integer()))
+    dict(ref=reference()))
 
 
 
-def item1to2(old):
+nonUpgradedItem2 = declareLegacyItem(
+    UpgradedItem.typeName, 2,
+    dict(ref=reference()))
+
+registerAttributeCopyingUpgrader(UpgradedItem, 1, 2)
+
+
+def item2to3(old):
     """
     Upgrade an nonUpgradedItem to UpgradedItem
     """
-    return old.upgradeVersion(UpgradedItem.typeName, 1, 2, dummy=old.dummy)
+    return old.upgradeVersion(UpgradedItem.typeName, 2, 3, ref=old.ref)
 
-registerUpgrader(item1to2, UpgradedItem.typeName, 1, 2)
+registerUpgrader(item2to3, UpgradedItem.typeName, 2, 3)

--- a/axiom/test/test_upgrading.py
+++ b/axiom/test/test_upgrading.py
@@ -260,7 +260,6 @@ class SwordUpgradeTest(SchemaUpgradeTest):
 
             loggedErrors = self.flushLoggedErrors(brokenapp.UpgradersAreBrokenHere)
             self.assertEqual(len(loggedErrors), 1)
-            originalError = loggedErrors[0]
 
             oldType = item.declareLegacyItem(
                 oldapp.Sword.typeName,

--- a/axiom/test/test_upgrading.py
+++ b/axiom/test/test_upgrading.py
@@ -239,7 +239,14 @@ class SwordUpgradeTest(SchemaUpgradeTest):
         Verify that if an exception is raised in an upgrader, the exception
         will be logged.
         """
-        playerID, swordID = self._testTwoObjectUpgrade()
+        choose(oldapp)
+        s = self.openStore()
+        swordID = oldapp.Sword(
+            store=s,
+            name=u'flaming vorpal doom',
+            hurtfulness=7).storeID
+        self.closeStore()
+
         choose(brokenapp)
         s = self.openStore()
         self.startStoreService()

